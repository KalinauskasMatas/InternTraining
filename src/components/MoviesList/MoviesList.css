.available-movies {
  background-color: var(--bright-background-color);
  margin: 14px 40px 106px 40px;
  border-radius: 20px;
}

.available-movies-header {
  margin: 0;
  padding: 16px 0 6px 0;
  font-weight: 400;
  letter-spacing: -1px;
  margin-bottom: 23px;
}

.available-movies table {
  padding: 0px 50px 30px 30px;
  width: 100%;
  padding-bottom: 68px;
  border-spacing: 0 10px;
}

.available-movies tr:nth-child(n+2) {
  background-color: var(--background-color);
}

.available-movies th {
  padding: 0 17px;
  text-align: left;
}

.available-movies td {
  padding: 8px 17px;
}

.available-movies td:not(.isInStock, .movie-rent) {
  width: 100px;
  text-align: left;
}


.available-movies td:first-child {
  border-top-left-radius: 18px;
  border-bottom-left-radius: 18px;
  width: 350px;
  padding-left: 17px;
}

.isInStock img {
  width: 20px;
}

.available-movies .movie-rent {
  background-color: var(--dark-button-color);
  color: var(--bright-text-color);
  padding: 10px 30px;
  cursor: pointer;
}

.available-movies .disabled-button {
  background: rgba(0, 0, 0, 0.4);
  cursor: not-allowed; 
}
