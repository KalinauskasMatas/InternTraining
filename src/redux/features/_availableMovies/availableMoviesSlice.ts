import { createSlice, PayloadAction } from "@reduxjs/toolkit";
// import axios from "axios";

import { MovieInterface } from "../../../interfaces";
import { RootState } from "../../store";

const initialState: MovieInterface[] = [];

const availableMoviesSlice = createSlice({
  name: "availableMovies",
  initialState,
  reducers: {
    returnMovie: (state, action: PayloadAction<string>) => {
      const movieIndex = state.findIndex(
        (movie: MovieInterface) => movie.name === action.payload
      );

      if (movieIndex === -1) return;
      state[movieIndex].stock++;
      localStorage.setItem("availableMovies", JSON.stringify(state));
    },
  },
});

export const { returnMovie } = availableMoviesSlice.actions;

// export const selectAvailableMovies = (state: RootState) =>
//   state.availableMovies;

export default availableMoviesSlice.reducer;
