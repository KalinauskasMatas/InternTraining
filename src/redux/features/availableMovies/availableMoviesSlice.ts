import { createSlice, PayloadAction } from "@reduxjs/toolkit";
// import axios from "axios";

import { MovieInterface } from "../../interfaces";
import { RootState } from "../../store";

import defaultData from "../../../defaultData.json";

// async function getMovies() {
//   try {
//     const response = await axios.get("./defaultData.json");
//     return response;
//   } catch (error) {
//     console.error(error);
//   }
// }

const initialState: MovieInterface[] =
  JSON.parse(localStorage.getItem("availableMovies") as string) || defaultData;

const availableMoviesSlice = createSlice({
  name: "availableMovies",
  initialState,
  reducers: {
    rentMovie: (state, action: PayloadAction<string>) => {
      const movieIndex = state.findIndex(
        (movie: MovieInterface) => movie.name === action.payload
      );
      if (movieIndex === -1) return;
      if (state[movieIndex].stock < 1) return;
      state[movieIndex].stock--;
    },
  },
});

export const { rentMovie } = availableMoviesSlice.actions;

export const selectAvailableMovies = (state: RootState) =>
  state.availableMovies;

export default availableMoviesSlice.reducer;
