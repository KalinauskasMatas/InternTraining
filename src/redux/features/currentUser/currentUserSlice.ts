import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "../../store";
import { UserState } from "../../interfaces";

const initialState: UserState =
  JSON.parse(localStorage.getItem("currUser") as string) || {};

const currentUserSlice = createSlice({
  name: "currentUser",
  initialState,
  reducers: {
    setUser: (state, action: PayloadAction<UserState>) => {
      state.fname = action.payload.fname;
      state.surname = action.payload.surname;
      state.email = action.payload.email;
      state.password = action.payload.password;
      localStorage.setItem("currUser", JSON.stringify(state));
    },
    clearUser: (state) => {
      localStorage.removeItem("currUser");
      state = { ...initialState };
      console.log(state);
    },
  },
});

export const { setUser, clearUser } = currentUserSlice.actions;

export const selectCurrUser = (state: RootState) => state.currentUser;

export default currentUserSlice.reducer;
