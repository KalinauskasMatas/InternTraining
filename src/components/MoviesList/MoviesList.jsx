import { useAppDispatch, useAppSelector } from "../../redux/hooks";

import { rentMovie } from "../../redux/features/availableMovies/availableMoviesSlice";

import "./MoviesList.css";

import checkIcon from "./assets/check.png";
import crossIcon from "./assets/cross.png";
import { updateCurrUser } from "../../redux/features/currentUser/currentUserSlice";
import { updateUser } from "../../redux/features/registeredUsers/registeredUsersSlice";

const MoviesList = () => {
  const currentUser = useAppSelector((state) => state.currentUser);
  const availableMovies = useAppSelector((state) => state.availableMovies);
  const dispatch = useAppDispatch();

  const rent = (movieName) => {
    const isMovieRented =
      currentUser.rentMovies.filter((movie) => movie.name === movieName)
        .length > 0;
    if (isMovieRented) return;

    const foundMovie = availableMovies.filter(
      (movie) => movie.name === movieName
    )[0];

    const newRentMovie = {
      name: movieName,
      genre: foundMovie.genre,
      time: 12,
      price: foundMovie.rentalPrice,
    };

    dispatch(rentMovie(movieName));

    dispatch(
      updateCurrUser({
        rentMovies: [...currentUser.rentMovies, newRentMovie],
      })
    );

    dispatch(
      updateUser({
        ...currentUser,
        rentMovies: [...currentUser.rentMovies, newRentMovie],
      })
    );
  };

  return (
    <section className="available-movies">
      <h2 className="available-movies-header">Available Movies</h2>
      <table id="available-movies-table">
        <tbody>
          <tr>
            <th>Name</th>
            <th>Genre</th>
            <th>Price for 12h</th>
            <th>Is in stock</th>
          </tr>
          {availableMovies.map((movie) => (
            <tr key={movie.name}>
              <td>{movie.name}</td>
              <td>{movie.genre}</td>
              <td>{movie.rentalPrice}$</td>
              <td className="isInStock">
                {movie.stock > 0 ? (
                  <img src={checkIcon} alt="Yes" />
                ) : (
                  <img src={crossIcon} alt="No" />
                )}
              </td>
              <td className="movie-rent" onClick={() => rent(movie.name)}>
                Rent
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </section>
  );
};

export default MoviesList;
