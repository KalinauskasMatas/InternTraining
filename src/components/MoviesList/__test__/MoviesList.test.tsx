import MoviesList from "../MoviesList";

import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";

import renderer from "react-test-renderer";

import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import availableMoviesSlice from "../../../redux/features/availableMovies/availableMoviesSlice";
import currentUserSlice from "../../../redux/features/currentUser/currentUserSlice";

const store = configureStore({
  reducer: {
    availableMovies: availableMoviesSlice,
    currentUser: currentUserSlice,
  },
});

afterEach(cleanup);

it("renders without crashing", () => {
  render(
    <Provider store={store}>
      <MoviesList />
    </Provider>
  );
});

it("matches snapshot", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <MoviesList />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
