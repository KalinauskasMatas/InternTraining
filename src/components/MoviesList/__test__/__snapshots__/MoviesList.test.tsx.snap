// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<section
  className="available-movies"
>
  <h2
    className="available-movies-header"
  >
    Available Movies
  </h2>
  <table
    id="available-movies-table"
  >
    <tbody>
      <tr>
        <th>
          Name
        </th>
        <th>
          Genre
        </th>
        <th>
          Price for 12h
        </th>
        <th>
          Is in stock
        </th>
      </tr>
      <tr>
        <td>
          The Godfather
        </td>
        <td>
          Drama
        </td>
        <td>
          5.99
          $
        </td>
        <td
          className="isInStock"
        >
          <img
            alt="Yes"
            src="check.png"
          />
        </td>
        <td
          className="movie-rent"
          onClick={[Function]}
        >
          Rent
        </td>
      </tr>
      <tr>
        <td>
          The Shawshank Redemption
        </td>
        <td>
          Drama
        </td>
        <td>
          4.99
          $
        </td>
        <td
          className="isInStock"
        >
          <img
            alt="Yes"
            src="check.png"
          />
        </td>
        <td
          className="movie-rent"
          onClick={[Function]}
        >
          Rent
        </td>
      </tr>
      <tr>
        <td>
          Raging Bull
        </td>
        <td>
          Biography
        </td>
        <td>
          4.69
          $
        </td>
        <td
          className="isInStock"
        >
          <img
            alt="Yes"
            src="check.png"
          />
        </td>
        <td
          className="movie-rent"
          onClick={[Function]}
        >
          Rent
        </td>
      </tr>
      <tr>
        <td>
          Schindler's List
        </td>
        <td>
          Biography
        </td>
        <td>
          5.49
          $
        </td>
        <td
          className="isInStock"
        >
          <img
            alt="Yes"
            src="check.png"
          />
        </td>
        <td
          className="movie-rent"
          onClick={[Function]}
        >
          Rent
        </td>
      </tr>
      <tr>
        <td>
          Casablanca
        </td>
        <td>
          Drama
        </td>
        <td>
          3.99
          $
        </td>
        <td
          className="isInStock"
        >
          <img
            alt="Yes"
            src="check.png"
          />
        </td>
        <td
          className="movie-rent"
          onClick={[Function]}
        >
          Rent
        </td>
      </tr>
      <tr>
        <td>
          Citizen Kane
        </td>
        <td>
          Mystery
        </td>
        <td>
          4.99
          $
        </td>
        <td
          className="isInStock"
        >
          <img
            alt="Yes"
            src="check.png"
          />
        </td>
        <td
          className="movie-rent"
          onClick={[Function]}
        >
          Rent
        </td>
      </tr>
      <tr>
        <td>
          Gone with the Wind
        </td>
        <td>
          Romance
        </td>
        <td>
          2.99
          $
        </td>
        <td
          className="isInStock"
        >
          <img
            alt="No"
            src="cross.png"
          />
        </td>
        <td
          className="movie-rent disabled-button"
          id="no-stockGone with the Wind"
        >
          Rent
          <div
            className="react-tooltip styles-module_tooltip__mnnfp styles-module_dark__xNqje"
            role="tooltip"
            style={
              Object {
                "backgroundColor": "rgba(0, 0, 0, 1)",
              }
            }
          >
            There is no stock left
            <div
              className="react-tooltip-arrow styles-module_arrow__K0L3T"
              style={Object {}}
            />
          </div>
        </td>
      </tr>
      <tr>
        <td>
          The Wizard of Oz
        </td>
        <td>
          Adventure
        </td>
        <td>
          5.99
          $
        </td>
        <td
          className="isInStock"
        >
          <img
            alt="Yes"
            src="check.png"
          />
        </td>
        <td
          className="movie-rent"
          onClick={[Function]}
        >
          Rent
        </td>
      </tr>
    </tbody>
  </table>
</section>
`;
